---
title: "Differential abundance with glmmTMB, MS Before vs After treatment, stool"
author: "Alba Troci"
date: "`r Sys.Date()`"
output:
   BiocStyle::html_document:
      toc: true
      df_print: paged
      self_contained: true
      code_download: true
      highlight: tango
editor_options:
  chunk_output_type: console
params:
  FIGPATH: "figures/"
  d.out: "./"
---

# Preparations {.tabset}

## Description

Set global options + Load libraries

## Global options

```{r style, results="asis", cache=FALSE, message = F}
# Set knit global options
library("knitr")
options(digits = 2, width = 80)
golden_ratio <- (1 + sqrt(5)) / 2
opts_chunk$set(echo = TRUE,
               tidy = FALSE,
               include = TRUE,
               fig.path = params$FIGPATH,
               dev=c("png",'pdf'),
               fig.height = 5,
               fig.width = 4 * golden_ratio,
               comment = '  ',
               dpi = 300,
               cache = FALSE)


# Pretty outputs
library("rmarkdown")
library("BiocStyle")
library("ggthemes") # add ggplot2 themes
library("ggplot2")
library("ggpubr")

# Set seed for reproducibility
set.seed(100)

# Set plotting theme
theme_set(theme_few(base_size = 14))

# Set output directory
d.out <- params$d.out
rm(params)
```


## Libraries
```{r libraries, message=FALSE}
library("magrittr")
library("ggplot2")
library("tidyverse")
library("readxl")
library("dada2")
library("phyloseq")
library("gridExtra")
library("vegan")
library("plyr")
library("scales")
library("reshape2")
library("DESeq2")
library("dplyr") #for data manipulation
library("msa") #multiple sequence alignment
library("ape") #to create phylogenetic tree
library("randomForest") 
library("caret") 
library("broom")
library("mlbench")
library("plotROC")
library("Matrix")
library("limma")
library("Glimma")
library("edgeR")
library("BiocParallel")
library("variancePartition")
library("statmod")
library("glmmTMB")
library("bbmle")
library("LaplacesDemon")
library("MuMIn")
library("DHARMa")
library("sjmisc")
library("sjstats")
library("lme4")
library("emmeans")


```



## Phyloseq object

```{r import data}
# Phyloseq object: Filtered ASV based from ST Samples
d.ps_ST1 <- "~/IKMB/projects/MS/github.codes.ms/ms.before.after/data"
ps_ST1 <- "ps.ms.stool.rds" %>%
  paste0(d.ps_ST1,"/", .) %>% 
  readRDS() 
ps_ST1

ps_ST2 <- subset_samples(ps_ST1, Ocrevus %in% "yes")
ps_ST2

tax_table(ps_ST2)[,9] <- taxa_names(ps_ST2)

```


```{r, message=FALSE}
#Add a column named Group (Before=TP0,TP1 and After=TP2,TP3)
sample_data(ps_ST2)$Group[sample_data(ps_ST2)$TimePoint %in%
                           c("TP1")] <- "Before"
sample_data(ps_ST2)$Group[sample_data(ps_ST2)$TimePoint %in%
                           c("TP2", "TP3", "TP4")] <- "After"
```

```{r}

ps_ST <- ps_ST2

```

# Prepare covariates for the model {.tabset}

## Description

1.Sample Table: 
Separate sample_data, ocrevus treatment only, in two groups: Before treatment (TP0,TP1) 
and After treatment (TP3,TP4)

2.Taxa Table: 
Subset new taxa table from otu_table(ps_ST) for the remaining samples. 

## Sample+Taxa tables 

**Subset Sample Table**
```{r}
#Covariates:Time, Treatment(ocrevus+no ocrevus)
#extract unique samples
dat <- data.frame(SampleID=sample_names(ps_ST),
                  Subject=sample_data(ps_ST)$SampleNr,
                  SampleID.1=sample_data(ps_ST)$SampleId,
                  TimePoint=sample_data(ps_ST)$TimePoint,
                  Group=sample_data(ps_ST)$Group,
                  Treat=sample_data(ps_ST)$Ocrevus,
                  Depth = sample_sums(ps_ST))

```


```{r}
#join 2 tables in one, vertically
t1 <- dat %>%
  
  #TimePoint:TP2=2, TP3=3
  mutate(Group=ifelse(Group=='Before','0',
                          ifelse(Group=='After','1',NA)))

t <- subset(t1, TimePoint %in% c("TP1", "TP4"))
                                         
#Subject as numbers
s <- gsub("[^[:digit:]]", "", t$Subject)
t$Subject <- s
tab <- data.frame(t)
tab

tab <- mutate_at(tab, vars(Group), as.factor)

```

Check is we have balanced or non balanced dataset regarding treatment (variable of interest)
```{r}
#we will compare treated.yes with treated.no, check the number respectively
gr.before <- subset(tab, Group == "0")
dim(gr.before)
gr.after <- subset(tab, Group == "1")
dim(gr.after)
```

**Subset ASV table**
```{r}
#Create taxa table with the remaining samples
tabS <- data.frame(SampleID=tab[,c('SampleID')])
taxa <- as.data.frame((otu_table(ps_ST)))
taxa$SampleID <- row.names(otu_table(ps_ST))
taxatable1 <- merge(tabS, taxa, by="SampleID")
row.names(taxatable1) <- taxatable1$SampleID
taxatable1$SampleID <- NULL
dim(taxatable1)
```


# Filter {.tabset}

## Description

I will filter the updated taxa table (taxatable1) from low abundant taxa.

1.Abundance filter:
Extract taxa (ASV here) that have total abundance at least 50 taxa per sample.

2.Prevalence filter:
Extract taxa that are present at least 5% in all samples.

## Abundance filter
```{r}
#ASV that have 0 total counts in all samples
table(colSums(taxatable1==0)== nrow(tab))

#delete ASV that have total abundance in each sample less than 50
taxatable2 <- taxatable1[,colSums(taxatable1) > 50] %>% na.omit()

```


## Prevalence filter
```{r}
#filter based prevalence
#count nr of elements not 0 in each column
a <- data.frame(prev = (colSums(taxatable2 != 0 )))

#subset taxa with prevalence at least 5% of all samples
b <- subset(a, prev >= 0.05 * nrow(tab))
taxatable <- taxatable2[, (colSums(taxatable2 != 0 ) > 0.05 * nrow(tab))] %>% na.omit()
```



**Subset taxonomy table**
Subset taxonomy table tax_table(ps_ST) to the remaining taxa after filtering

```{r}
bb <- colnames(taxatable)
aa <- tax_table(ps_ST)[,c(1,2,3,4,5,6,7,9)]
taxonomy <- subset_taxa(aa, rownames(aa) %in% bb)
dim(taxonomy)
dim(taxatable)

t <- data.frame(taxonomy)
```


## Design

```{r}
#Prepare design matrix
Group <- factor(paste(tab$Group ))
design = model.matrix(~0+ Group)
design
tabt <- cbind(tab, design)
colnames(design) <- levels(Group)

#check if we have column with all zeros and delete it
all.zero <- apply(design, 2, function(x) all(x==0))
all.zero

#we need contrasts.arg for glmmTMB model 
contrasts.arg=list(Group=contrasts(as.factor(tabt$Group)))  
contrasts.arg

```


# Fit glmmTMB

I will fit glmmTMB (family nbinom2) for each ASV in taxatable:

* Create covariate matrix, X.

* Fit glmmTMB(Abundance ~ Group + (1|Subject)).

* Compute "logFoldChange" and "baseMean".

```{r, warning=FALSE, message=FALSE}
#Create a matrix to hold taxa
fail <- list()
warn <- list()
taxa_all <- colnames(taxatable)
data_out <- data.frame()
counter=0
for (tax in taxa_all){
  counter <- counter +1
  
  #Covariate matrix
  ab <- data.frame(Abundance= taxatable[[tax]], SampleID = rownames(taxatable))

  X <- merge(ab, tab, by = "SampleID")

  #cat(tax,'\n')
  data_out[counter,"taxon"] <- tax
  data_out[counter,"Abundance"] <- sum(X$Abundance)
  data_out[counter,"Prevalence"] <- length(which(X$Abundance != 0))
  
  tryCatch({
    #Fit model 
    fit_nbinom2 <- glmmTMB(Abundance~Group + (1|Subject)
                              +offset(log(Depth)),
                              data=X,
                              #ZI parameter applying to all observations
                              ziformula=~1, 
                              contrasts = contrasts.arg,
                              family=nbinom2)
  
    #Extract results  
    res_anova <- as.matrix(glmmTMB:::Anova.glmmTMB(fit_nbinom2, 
                                                   type = c("II"))) 
  
    data_out[counter,"Chisq"] <- res_anova[1,1]
    data_out[counter,"df"]  <- res_anova[1,2]
    data_out[counter,"P"]  <- res_anova[1,3]
  
    #Extract logFoldChange and baseMean
    emm <- as.matrix(emmeans(fit_nbinom2, poly ~ Group))
  
    
    #marginal means for cluster1(treat), cluster2(no treat)
    emm1 <- as.data.frame(emm[[1]])
    baseMean <- (emm1$emmean[1]+emm1$emmean[2])/2
    
    #logFoldChange
    emm2 <- as.data.frame(emm[[2]])
    logFoldChange <- emm2$estimate
    
    #Update result table
    data_out[counter,"baseMean"]  <- baseMean
    data_out[counter, "logFoldChange"] <- logFoldChange
    
  
  },error = function(c) { 
      #print(c$message)
      fail[tax] <<- 2},
   warning = function(w){
     #print(w$message)
     warn[tax] <<- 1
     })
}

```


```{r}
#take out ASV with warning 
data_clean1 <- subset(data_out, !(taxon %in% names(warn)))
data_clean <- subset(data_clean1, baseMean > 0)
data_clean
```


## Distribution of p values
```{r, warning=FALSE}
hist(data_clean$P, main="P-values:Treat",xlab="Treat") 
```


## Adjusted pvalues
```{r, warning=FALSE}
data_clean$P.adj.bon <- p.adjust(data_clean$P,"bon")
data_clean$P.adj.fdr <- p.adjust(data_clean$P,"fdr")
data_clean$P.adj.holm <- p.adjust(data_clean$P,"holm")

res <- data.frame(ASV=data_clean$taxon,
                  Abundance = data_clean$Abundance,
                  Prevalence = data_clean$Prevalence,
                  log2FoldChange=data_clean$logFoldChange,
                  baseMean=data_clean$baseMean,
                  Chisq=data_clean$Chisq,
                  df=data_clean$df,
                  P=data_clean$P, 
                  padj=data_clean$P.adj.holm)

#Result table
res_order <- res[order(res$padj),] %>% na.omit()
rownames(res_order) <- res_order$ASV
```

```{r}
#taxa with p value <0.05
sig_pvalue <- res_order[(res_order$padj ) <0.05 , ] %>% 
  na.omit()

#print only significant abundant taxa
sig_pvalue

```


## Plot significant taxa

Assign taxonomy to ASV that have significant p value:
```{r, warning=FALSE}

treat_p_tax <- merge(sig_pvalue, taxonomy, by= "ASV")
treat_p_tax
```

ASVs that were significantly different between the two groups (Before:After treatment with ocrevus):
```{r, message=FALSE, warning=FALSE}
theme_set(theme_bw())
sigtabgen <- treat_p_tax
sigtabgen = subset(sigtabgen, !is.na(Genus))
# Phylum order
x = tapply(sigtabgen$log2FoldChange, sigtabgen$Phylum, function(x) max(x))
x = sort(x, TRUE)
sigtabgen$Phylum = factor(as.character(sigtabgen$Phylum), levels=names(x))
# Genus order
x = tapply(sigtabgen$log2FoldChange, sigtabgen$Genus, function(x) max(x))
x = sort(x, TRUE)
sigtabgen$Genus = factor(as.character(sigtabgen$Genus), levels=names(x))
fig <- ggplot(sigtabgen, aes(y=Genus, x=log2FoldChange, color=Phylum)) + 
  geom_vline(xintercept = 0.0, color = "gray", size = 0.5) +
  geom_point(aes(size=(sigtabgen$baseMean)))+
    scale_size(range = c(1,3))+
  labs( size = "baseMean")+
  ggtitle("MS-O_TP1                                               MS-O_TP4")+
  theme(axis.text.x = element_text(angle = -90, hjust = 0, vjust=0.5), 
        text = element_text(size = 8),
        plot.title = element_text(size = 6)) +
    scale_color_brewer(palette = "Dark2")
  
fig

fig.ms.oc.tp1.tp4 <- fig

ggsave("ms.oc.tp1.tp4.png")
```



# save output

```{r}
export(treat_p_tax, "MS.stool.diff.abundance.oc.TP1.TP4.xlsx")
```

